### PR TITLE
[api] Get Pending Action by eth_getTransactionByHash

### PR DESCRIPTION
**Pull request by envestcc**
# Description
`eth_getTransationByHash` now support confirmed or pending actions. If action is pending, the `blockHash` and `blockNumber` is zero

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: api/web3server_marshal.go [hunk 1]
```diff
@@ -222,13 +222,25 @@ func (obj *getBlockResult) MarshalJSON() ([]byte, error) {
 }
 
 func (obj *getTransactionResult) MarshalJSON() ([]byte, error) {
```
Comment (dustinxie) [inline api/web3server_marshal.go:224]: as we chatted, add
```
func (obj *getTransactionResult) Construct(blkHash, selp, receipt)
```

Comment (envestcc) [reply api/web3server_marshal.go:263]: updated

**Comment by Liuhaai** [comment]
typo `eth_getTransactionByHash`

File: api/web3server_marshal.go [hunk 2]
```diff
@@ -221,14 +222,64 @@ func (obj *getBlockResult) MarshalJSON() ([]byte, error) {
 	})
 }
 
+func newGetTransactionResult(
```
Comment (Liuhaai) [inline api/web3server_marshal.go:225]: the funcs in this file are to implement `MarshalJSON()` interface

File: api/web3server.go [hunk 3]
```diff
@@ -598,7 +605,7 @@ func (svr *web3Handler) getTransactionByHash(in *gjson.Result) (interface{}, err
 		}
 		return nil, err
 	}
-	return svr.getTransactionFromActionInfo(blkHash, selp, receipt)
+	return svr.assembleConfirmedTransaction(blkHash, selp, receipt)
```
Comment (dustinxie) [inline api/web3server.go:608]: from L588
```
if err == nil {
	return svr.assembleConfirmedTransaction(blkHash, selp, receipt)
}
if errors.Cause(err) == ErrNotFound {
	selp, err = svr.coreService.PendingActionByActionHash(actHash)
}
if err == nil {
	return svr.assemblePendingTransaction(selp)
}
if errors.Cause(err) == ErrNotFound {
	return nil, nil
}
return nil, err
```

File: api/web3server_marshal.go [hunk 4]
```diff
@@ -221,14 +222,64 @@ func (obj *getBlockResult) MarshalJSON() ([]byte, error) {
...
+	if err != nil {
+		return nil, err
+	}
+	var to *string
+	if ethTx.To() != nil {
+		tmp := ethTx.To().String()
+		to = &tmp
+	}
+
+	signer, err := action.NewEthSigner(iotextypes.Encoding(selp.Encoding()), chainID)
```
Comment (dustinxie) [inline api/web3server_marshal.go:246]: should be `coreService.EVMNetworkID())`

File: api/web3server_marshal.go [hunk 5]
```diff
@@ -221,14 +222,64 @@ func (obj *getBlockResult) MarshalJSON() ([]byte, error) {
 	})
 }
 
+func newGetTransactionResult(
+	blkHash hash.Hash256,
+	selp action.SealedEnvelope,
+	receipt *action.Receipt,
+	chainID uint32,
```
Comment (dustinxie) [inline api/web3server_marshal.go:229]: evmChainID

File: api/web3server_marshal.go (hunk 2 above)
Comment (envestcc) [reply api/web3server_marshal.go:225]: move to web3server_utils.go?

File: api/web3server.go (hunk 3 above)
Comment (envestcc) [reply api/web3server.go]: updated to reduce code nesting

File: api/coreservice.go [hunk 6]
```diff
@@ -108,6 +108,8 @@ type (
 		ActionsByAddress(addr address.Address, start uint64, count uint64) ([]*iotexapi.ActionInfo, error)
 		// ActionByActionHash returns action by action hash
 		ActionByActionHash(h hash.Hash256) (action.SealedEnvelope, hash.Hash256, uint64, uint32, error)
+		// PendingActionByActionHash returns action by action hash
+		PendingActionByActionHash(h hash.Hash256) (action.SealedEnvelope, error)
```
Comment (Liuhaai) [inline api/coreservice.go:112]: utilize `Action(actionHash string, checkPending bool) (*iotexapi.ActionInfo, error)`?

File: api/web3server.go [hunk 7]
```diff
@@ -585,20 +585,25 @@ func (svr *web3Handler) getTransactionByHash(in *gjson.Result) (interface{}, err
...
-		if errors.Cause(err) == ErrNotFound {
+	if err == nil {
+		receipt, err := svr.coreService.ReceiptByActionHash(actHash)
+		if err != nil && errors.Cause(err) == ErrNotFound {
 			return nil, nil
+		} else if err != nil {
+			return nil, err
 		}
-		return nil, err
+		return svr.assembleConfirmedTransaction(blkHash, selp, receipt)
```
Comment (dustinxie) [inline api/web3server.go:595]:
```
if err == nil {
	return svr.assembleConfirmedTransaction(blkHash, selp, receipt)
}
if errors.Cause(err) == ErrNotFound {
	return nil, nil
}
return nil, err
```

File: api/web3server.go [hunk 8]
```diff
@@ -585,20 +585,25 @@ func (svr *web3Handler) getTransactionByHash(in *gjson.Result) (interface{}, err
...
-		if errors.Cause(err) == ErrNotFound {
+	if errors.Cause(err) == ErrNotFound {
+		selp, err = svr.coreService.PendingActionByActionHash(actHash)
+		if err != nil && errors.Cause(err) == ErrNotFound {
 			return nil, nil
+		} else if err != nil {
+			return nil, err
 		}
-		return nil, err
+		return svr.assemblePendingTransaction(selp)
```
Comment (dustinxie) [inline api/web3server.go:604]: same as above

**Review by dustinxie — APPROVED** [review]

File: api/coreservice.go (hunk 6 above)
Comment (envestcc) [reply api/coreservice.go:112]: A great suggestion. However, I noticed that the coreservice provides two interfaces with duplicate functionality:
```
Action(actionHash string, checkPending bool) (*iotexapi.ActionInfo, error)
ActionByActionHash(h hash.Hash256) (action.SealedEnvelope, hash.Hash256, uint64, uint32, error)
```

The returned action structures are different. The first is currently used for gRPC, while the second is used for web3. Therefore, I added an interface specifically for web3 to query pending actions, to accommodate web3's processing logic:
```
PendingActionByActionHash(h hash.Hash256) (action.SealedEnvelope, error)
```

Comment (envestcc) [reply api/coreservice.go:112]: However, it seems that coreservice should provide a unified interface for both gRPC and web3. We can merge into a unified function in a separate PR.

File: api/web3server.go (hunk 7 above)
Comment (envestcc) [reply api/web3server.go]: updated

File: api/coreservice.go (hunk 6 above)
Comment (dustinxie) [reply api/coreservice.go:112]: yes, let's do that in another PR

File: api/web3server_marshal.go [hunk 9]
```diff
@@ -245,11 +257,11 @@ func (obj *getTransactionResult) MarshalJSON() ([]byte, error) {
 		S                string  `json:"s"`
 		V                string  `json:"v"`
 	}{
-		Hash:             "0x" + hex.EncodeToString(obj.receipt.ActionHash[:]),
+		Hash:             "0x" + hex.EncodeToString(txHash),
```
Comment (Liuhaai) [inline api/web3server_marshal.go:260]: follow the specifications https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_gettransactionbyhash

**Review by Liuhaai — CHANGES_REQUESTED** [review]

File: api/web3server_marshal.go (hunk 9 above)
Comment (envestcc) [reply api/web3server_marshal.go:260]: Thx for the correction: The three fields (i.e. `blockHash`, `blockNumber` and `transactionIndex`) should be `null` if transaction is pending.
Here is example call from eth:
```
$ curl -X POST -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"eth_getTransactionByHash","params":["0x7443280c0df6c69fb643ffde0534508dcefa688906ff0c2dc45041548b6418df"],"id":0}' https://ethereum.iotex.one/v1/mainnet

{
   "id" : 0,
   "jsonrpc" : "2.0",
   "result" : {
      "blockHash" : null,
      "blockNumber" : null,
      "chainId" : "0x1",
      "from" : "0x2e05a304d3040f1399c8c20d2a9f659ae7521058",
      "gas" : "0x1fbd0",
      "gasPrice" : "0x304fed86c",
      "hash" : "0x7443280c0df6c69fb643ffde0534508dcefa688906ff0c2dc45041548b6418df",
      "input" : "0xa5e387511b44474377bfe2d5a56bf69a2e279de9b29c54b7ec78aa0353be740c81a6c425",
      "nonce" : "0x71d7d",
      "r" : "0xae22f44be8e3cf52504e3cba3b8852774021fe4588535df796a0170707a61732",
      "s" : "0x70be719fc2b35eabc302695ca7d0e020fdcdf5104ec14eb9a13913fec2653cff",
      "to" : "0x5be1de8021cc883456fd11dc5cd3806dbc48d304",
      "transactionIndex" : null,
      "type" : "0x0",
      "v" : "0x26",
      "value" : "0x0"
   }
}
```

File: api/coreservice.go (hunk 6 above)
Comment (Liuhaai) [reply api/coreservice.go:112]: add TODO for the improvement

**Review by Liuhaai — APPROVED** [review]

**Pull request merged**